tests/csapi/device_lists_test.go

```diff
@@ -6,61 +6,315 @@ import (
+			client.SyncReq{Since: aliceNextBatch},
+			syncDeviceListsHas("changed", bob.UserID),
+		)
+		mustQueryKeys(t, alice, bob.UserID, checkBobKeys)
+	}
+
+	// testOtherUserLeave tests another user leaving a room Alice is in.
+	testOtherUserLeave := func(t *testing.T, deployment *docker.Deployment, hsName string, otherHSName string) {
+		alice := deployment.RegisterUser(t, hsName, generateLocalpart("alice"), "password", false)
+		bob := deployment.RegisterUser(t, otherHSName, generateLocalpart("bob"), "password", false)
+		checkBobKeys := uploadNewKeys(t, bob)
+
+		roomID := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+
+		// Bob joins the room
+		bob.JoinRoom(t, roomID, []string{hsName})
+		bobNextBatch := bob.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(bob.UserID, roomID))
+
+		// Alice performs an initial sync
+		alice.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(bob.UserID, roomID))
+		mustQueryKeys(t, alice, bob.UserID, checkBobKeys)
+		// Some homeservers (Synapse) may emit another `changed` update after querying keys.
+		_, aliceNextBatch := alice.MustSync(t, client.SyncReq{TimeoutMillis: "0"})
+
+		// Bob leaves the room
+		bob.LeaveRoom(t, roomID)
+		bob.MustSyncUntil(t, client.SyncReq{Since: bobNextBatch}, client.SyncLeftFrom(bob.UserID, roomID))
+
+		// Check that Alice is notified
+		alice.MustSyncUntil(
+			t,
+			client.SyncReq{Since: aliceNextBatch},
+			syncDeviceListsHas("left", bob.UserID),
+		)
+
+		// Both homeservers think Bob has left now
+		// Bob then updates their device list
+		// Alice's homeserver is not expected to get the device list update and must not return a
+		// cached device list for Bob.
+		checkBobKeys = uploadNewKeys(t, bob)
+		mustQueryKeys(t, alice, bob.UserID, checkBobKeys)
+	}
+
+	// testLeave tests Alice leaving a room another user is in.
+	testLeave := func(t *testing.T, deployment *docker.Deployment, hsName string, otherHSName string) {
+		alice := deployment.RegisterUser(t, hsName, generateLocalpart("alice"), "password", false)
+		bob := deployment.RegisterUser(t, otherHSName, generateLocalpart("bob"), "password", false)
+		checkBobKeys := uploadNewKeys(t, bob)
+
+		roomID := bob.CreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+
+		// Alice joins the room
+		alice.JoinRoom(t, roomID, []string{otherHSName})
+		bobNextBatch := bob.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(alice.UserID, roomID))
+
+		// Alice performs an initial sync
+		alice.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(alice.UserID, roomID))
+		mustQueryKeys(t, alice, bob.UserID, checkBobKeys)
+		// Some homeservers (Synapse) may emit another `changed` update after querying keys.
+		_, aliceNextBatch := alice.MustSync(t, client.SyncReq{TimeoutMillis: "0"})
+
+		// Alice leaves the room
+		alice.LeaveRoom(t, roomID)
+		bob.MustSyncUntil(t, client.SyncReq{Since: bobNextBatch}, client.SyncLeftFrom(alice.UserID, roomID))
+
+		// Check that Alice is notified
```

Comment:
notified of what? I think something like this(??)

```suggestion
		// Check that Alice is notified that she will no longer receive updates about Bob's devices.
```

Reply:
That's better, thanks. Updated all 4 places with similar wording.